main.go

```diff
@@ -205,23 +206,24 @@ func action(w http.ResponseWriter, r *http.Request) {
 		}
 	case "copy":
 		if len(names) > 0 {
-			http.SetCookie(w, &http.Cookie{Name: "cpdir", Value: p})
+			http.SetCookie(w, &http.Cookie{Name: "cpdir", Value: url.QueryEscape(p)})
 			encodedNames := strings.Join(names, "!$!")
-			http.SetCookie(w, &http.Cookie{Name: "cpitems", Value: string(encodedNames)})
+			http.SetCookie(w, &http.Cookie{Name: "cpitems", Value: url.QueryEscape(encodedNames)})
 			http.SetCookie(w, &http.Cookie{Name: "delorigin", Value: "false"})
 		}
 		http.Redirect(w, r, p, http.StatusMovedPermanently)
 	case "cut":
 		if len(names) > 0 {
-			http.SetCookie(w, &http.Cookie{Name: "cpdir", Value: p})
+			http.SetCookie(w, &http.Cookie{Name: "cpdir", Value: url.QueryEscape(p)})
 			encodedNames := strings.Join(names, "!$!")
-			http.SetCookie(w, &http.Cookie{Name: "cpitems", Value: string(encodedNames)})
+			http.SetCookie(w, &http.Cookie{Name: "cpitems", Value: url.QueryEscape(encodedNames)})
 			http.SetCookie(w, &http.Cookie{Name: "delorigin", Value: "true"})
 		}
 		http.Redirect(w, r, p, http.StatusMovedPermanently)
 	case "paste":
 		var cpdir *http.Cookie
 		cpdir, err = r.Cookie("cpdir")
+		cpdirValue, err := url.QueryUnescape(cpdir.Value)
```

Comment:
If line 255 gets an error, cpdir will be null and line 266 will panic, causing the whole server to crash.